pallets/loan/src/functions.rs

```diff
@@ -193,89 +195,102 @@ impl<T: Config> Pallet<T> {
 				// update loan status
 				loan_info.status = LoanStatus::Closed;
 				*maybe_loan_info = Some(loan_info);
-				Ok(asset)
+				Ok((asset, written_off))
 			},
 		)
 	}
 
+	// tries to borrow some amount on a loan that is active.
+	// returns a bool indicating if this is the first borrow or not.
 	pub(crate) fn borrow_amount(
 		pool_id: PoolIdOf<T>,
 		loan_id: T::LoanId,
 		owner: T::AccountId,
 		amount: T::Amount,
-	) -> DispatchResult {
+	) -> Result<bool, DispatchError> {
 		// ensure owner is the loan owner
 		Self::check_loan_owner(pool_id, loan_id, owner.clone())?;
 
-		LoanInfo::<T>::try_mutate(pool_id, loan_id, |maybe_loan_info| -> DispatchResult {
-			let mut loan_info = maybe_loan_info.take().ok_or(Error::<T>::ErrMissingLoan)?;
+		LoanInfo::<T>::try_mutate(
+			pool_id,
+			loan_id,
+			|maybe_loan_info| -> Result<bool, DispatchError> {
+				let mut loan_info = maybe_loan_info.take().ok_or(Error::<T>::ErrMissingLoan)?;
 
-			// ensure loan is active
-			ensure!(
-				loan_info.status == LoanStatus::Active,
-				Error::<T>::ErrLoanNotActive
-			);
+				// ensure loan is active
+				ensure!(
+					loan_info.status == LoanStatus::Active,
+					Error::<T>::ErrLoanNotActive
+				);
 
-			// ensure maturity date has not passed if the loan has a maturity date
-			let now: u64 = Self::time_now()?;
-			let valid = match loan_info.loan_type.maturity_date() {
-				// loan has a maturity date
-				Some(md) => md > now,
-				// no maturity date, so continue as is
-				None => true,
-			};
-			ensure!(valid, Error::<T>::ErrLoanMaturityDatePassed);
+				// ensure loan is not written off
+				ensure!(
+					loan_info.write_off_index.is_none(),
+					Error::<T>::ErrLoanWrittenOffByAdmin
+				);
 
-			// ensure borrow amount is positive
-			ensure!(amount.is_positive(), Error::<T>::ErrLoanValueInvalid);
+				// ensure maturity date has not passed if the loan has a maturity date
+				let now: u64 = Self::time_now()?;
+				let valid = match loan_info.loan_type.maturity_date() {
+					// loan has a maturity date
```

Comment:
Personally these comments inside the `match` don't add anything to the (clear) code so I would remove them

Reply:
fair!